flowinfra: avoid locking mutex in ScheduleFlow happy case

asubiotto opened the pull request:
The point of locking the mutex is to check whether a flow the scheduler is
about to run would push the number of running flows over the limit. This has
shown to be a source of significant mutex contention in TPC-E (see #50022).

This commit implements that check using atomics, which is cheaper than locking
the mutex.

Release note (performance improvement): unnecessary mutex contention observed
in heavy read workloads has been removed.

Closes #50022

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51055)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-:-MBdrHJg3DARyx8rTivd:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @nvanbenschoten)

___
*[pkg/sql/flowinfra/flow_scheduler_test.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-MBdqXB02ErNlrh7Komt:-MBdqXB14XwmbeQuC3-3:bbdw0xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73a4dc11ded9792f2d3827f7423744eed65cc77/pkg/sql/flowinfra/flow_scheduler_test.go#L42)):*
> ```Go
> 	doneCb func()
> }
> ```

nit: maybe add `var _ Flow = &mockFlow` assertion.
___
*[pkg/sql/flowinfra/flow_scheduler_test.go, line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-MBdr3r36t0Nkz1F66XF:-MBdr3r36t0Nkz1F66XG:bc2ifc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73a4dc11ded9792f2d3827f7423744eed65cc77/pkg/sql/flowinfra/flow_scheduler_test.go#L131)):*
> ```Go
> 	close(flow1.doneCh)
> 	// Now that flow1 has finished, flow2 should be run.
> 	<-flow2.runCh
> ```

What's the goal of receiving from the channel here? Making sure that `flow2.runCh` is closed?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51055)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/flowinfra/flow_scheduler.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-MBdu0UiC4GEc8u1DXxt:-MBdu0UiC4GEc8u1DXxu:bac39vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73a4dc11ded9792f2d3827f7423744eed65cc77/pkg/sql/flowinfra/flow_scheduler.go#L129)):*
> ```Go
> 	return fs.stopper.RunTaskWithErr(
> 		ctx, "flowinfra.FlowScheduler: scheduling flow", func(ctx context.Context) error {
> 			fs.mu.Lock()
> ```

I must be missing something, but aren't we still grabbing an exclusive lock and holding it through `Flow.Start` here?


<!-- Sent from Reviewable.io -->

Commit:
flowinfra: avoid locking mutex in ScheduleFlow happy case

The point of locking the mutex is to check whether a flow the scheduler is
about to run would push the number of running flows over the limit. This has
shown to be a source of significant mutex contention in TPC-E (see #50022).

This commit implements that check using atomics, which is cheaper than locking
the mutex.

Release note (performance improvement): unnecessary mutex contention observed
in heavy read workloads has been removed.

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51055)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten and @yuzefovich)

___
*[pkg/sql/flowinfra/flow_scheduler.go, line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-MBdu0UiC4GEc8u1DXxt:-MBhRyCq1sLNyWMkQk6Z:bx3h9k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73a4dc11ded9792f2d3827f7423744eed65cc77/pkg/sql/flowinfra/flow_scheduler.go#L129)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I must be missing something, but aren't we still grabbing an exclusive lock and holding it through `Flow.Start` here?
</blockquote></details>

:man_facepalming: forgot to change this.
___
*[pkg/sql/flowinfra/flow_scheduler_test.go, line 131 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-MBdr3r36t0Nkz1F66XF:-MBhnDX03cBoGepZs_Mr:b-2ikff2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b73a4dc11ded9792f2d3827f7423744eed65cc77/pkg/sql/flowinfra/flow_scheduler_test.go#L131)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

What's the goal of receiving from the channel here? Making sure that `flow2.runCh` is closed?
</blockquote></details>

Yeah, which indicates the `flow2.Run` or `flow2.Start` has been called (comment on `runCh` field).


<!-- Sent from Reviewable.io -->

asubiotto commented:
Friendly ping

yuzefovich reviewed (approved):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51055)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51055#-:-ME37Y-m-gkhSilhR8xF:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

asubiotto commented:
TFTRs!

bors r=yuzefovich,nvanbenschoten
